### ISSUE TITLE
Jittery and broken SDL_WarpMouseInWindow()

### ISSUE THREAD

**Issue by fgenesis** [libsdl-org/SDL#5609]
TL;DR 82793ac279d19b broke SDL_WarpMouse() and it's been semi-broken in various forms since then.

System specs: win10-x64/msvc19. Affected game: [Aquaria](https://github.com/AquariaOSE/Aquaria), tested on build-refactor branch but should apply to master as well.

Note that Aquaria grabs the mouse on startup in case that is relevant.
Did git a bisect, found a few guilty commits.

To reproduce: Enter any map, press RMB to start singing, spin the song circle like a maniac.
(Note that when singing, Aquaria calls SDL_WarpMouse(x, y) to keep the mouse in a circle around the screen center, with (x, y) equal or very close to the current mouse position; that might be a factor?)

Expected: The cursor should rotate smoothly in a circle. That is fine pre- 82793ac279d19b and has worked like this even with SDL-1.2.

Actual, between [ 82793ac279d19b ... 31f8c3ef4409a93 ): The game freezes hard and deadlocks the mouse (infinite MouseMotion events since the event handler calls SDL_WarpMouseInWindow() ).
Actual, between [ 31f8c3ef4409a93 ... 3318590796744 ): The game freezes but Alt+Tab gets out of it somehow and it functions normally until we start singing again.
Actual, between [3318590796744 ... HEAD ]: The mouse is stuck-ish and jerky, as if the sensor was covered in grime or the mouse balls super worn down from years of prior use on sandpaper while now frantically trying to operate that poor mouse in a puddle of beer.

I've somewhat alleviated the problem in https://github.com/AquariaOSE/Aquaria/commit/764d106d507ac07c5ced97abe5e24ea1cd716529 but decided to open this issue because it's still sort of broken in a not backwards compatible way.

Perhaps the most curious part is that the right half of the song circle functions normally. The mouse movement becomes jerky only on the outer left side of the circle.

See also #5457, #5569

**Comment by Photosounder** [comment]
Wow yes I just upgraded from 2.0.16 to 2.0.22 and it broke my programs in a major way. Had to roll back to 2.0.20.

**Comment by slouken** [comment]
> Wow yes I just upgraded from 2.0.16 to 2.0.22 and it broke my programs in a major way. Had to roll back to 2.0.20.

Can you provide repro steps with your program and explain what you’re seeing?

Thanks!

**Comment by slouken** [comment]
@fgenesis, can you send me e-mail at slouken@libsdl.org? I'm trying to build and run Aquaria and am running into issues.

Thanks!

**Comment by Photosounder** [comment]
> Can you provide repro steps with your program and explain what you’re seeing?
> 
> Thanks!

I made a project to test it, it has a GUI with a log of events and a test control. You can directly try the Window binary in the releases. It will show you the mouse motion events. The 2.0.22 bug creates mouse events in contradictory directions. More details in the readme.
https://github.com/Photosounder/SDL-event-logger

**Comment by slouken** [comment]
Thanks for the test!
@fgenesis, @Photosounder, can you try the commit https://github.com/libsdl-org/SDL/commit/b44241ab2f626965912d419f6406b9336fe7969c and see if that fixes things for you?

**Comment by Photosounder** [comment]
Thanks, that fixes it for me.
